File: sortsearch/arraydivision.py
from threading import Thread
from sys import stdin


def split(li, mid):
    pieces = 1
    tempSum = 0
    for i in li:
        if tempSum + i > mid:
            pieces += 1
            tempSum = i
        else:
            tempSum += i
    return pieces


def main():
    n, k = map(int, stdin.readline().split())
    li = [int(c) for c in stdin.readline().split()]

    mx = 0
    s = 0
    for i in range(n):
        if li[i] > mx:
            mx = li[i]
        s += li[i]

    lo = mx
    hi = s

    while lo < hi:
        mid = (lo + hi) // 2
        pieces = split(li, mid)
        if pieces > k:
            lo = mid + 1
        else:
            hi = mid

    print(lo)


Thread(target=main).start()

File: sortsearch/arraydivision.cc
#include<bits/stdc++.h>
using namespace std;
typedef long long int ll;

ll split(ll li[], ll mid, ll n) {
    ll pieces = 1;
    ll tempSum = 0;
    for (ll i = 0; i < n; i++) {
        if (tempSum + li[i] > mid) {
            pieces += 1;
            tempSum = li[i];
        }
        else {
            tempSum += li[i];
        }
    }
    return pieces;
}

int main() {
    ll n, k, mx, s;
    cin >> n >> k;
    ll li[n];
    mx = 0;
    s = 0;
    for (ll i = 0; i < n; i++) {
        cin >> li[i];
        s += li[i];
        if (li[i] > mx)
            mx = li[i];
    }
    ll lo = mx;
    ll hi = s;

    while (lo < hi){
        ll mid = (lo + hi) / 2;
        ll pieces = split(li, mid, n);
        if (pieces > k) {
            lo = mid + 1;
        }
        else {
            hi = mid;
        }
    }
    cout << lo;

    return 0;
}